### PR TITLE
Bug 1160982 - [Stingray][EPG] Fetch program information from TV API a…

### PR DESCRIPTION
**Pull request by sean2449**
…nd connect with view
- Add channel manager to connect TV API
- Add program mocks
- Display channels
- Display programs and timeline
- Update clock time

### CODE REVIEW

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2ac043370e25f3e66b80e1a9b19d2f0e3b9fbd9c

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=463f4c594c5da7c4994326a2485d30eb7dcb2dda

File: tv_apps/tv-deck/js/mock_tv_service.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/* global TVSource, TVChannel */
+/* global TVSource, TVChannel, TVProgram */
```
Comment (huchengtw-moz) [inline tv_apps/tv-deck/js/mock_tv_service.js]: You had also changed this file. Is this what you want?

File: tv_apps/tv-epg/js/channel_manager.js [hunk 2]
```diff
@@ -0,0 +1,244 @@
+/* global evt */
+
+'use strict';
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/channel_manager.js]: It looks like this file is totally the same as channel_manager in TV. Should we put them at `shared/smart-screen/tv` folder?

File: tv_apps/tv-epg/js/epg.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
+/* global EPGController */
+
+'use strict';
+
+(function(exports) {
+
+  function EPG() {
+    this.timelineUnit = 30 * 60000;
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg.js]: We should write a comment to express the unit is 30min.

File: tv_apps/tv-epg/js/epg.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    this.timeOffset = Math.floor(Date.now() / this.timelineUnit);
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg.js]: Is time offset the timeSlot?

File: tv_apps/tv-epg/js/epg.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    this.timeWidth = 4;
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg.js]: What's timeWidth?

File: tv_apps/tv-epg/js/epg_controller.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+      var fetchChannelCount = 0;
+      var channelItems = this.channelManager.getSource().channels;
+      channelItems.forEach(function(channelItem, index) {
+        channelItem.channel.getPrograms({
+          startTime: startTime * this.timelineUnit,
+          duration: duration * this.timelineUnit
+        }).then(function(programs) {
+          this._addPrograms(index, programs);
+          fetchChannelCount++;
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg_controller.js]: We should use fetchedChannelCount

File: tv_apps/tv-epg/js/epg_controller.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+      endTime = startTime + Math.ceil(program.duration / this.timelineUnit);
+      endTime = Math.min(endTime, this.timelineOffset + this.timeslotCount);
+      for(time = startTime; time < endTime; time++) {
+        column = time - this.timelineOffset;
+        // filter out non-existing timeline
+        if (this.programTable[column]) {
+          // only the first timeslot within the interval will be displayed
+          if (isVisible) {
+            this.fire('addProgram', {
+              title: new Date(program.startTime).getHours() + ':' +
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg_controller.js]: Please use l10n_date.

File: tv_apps/tv-epg/js/epg.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    this.timeElement.textContent = timeString.time;
+
+    var clockUpdateRate = Math.ceil(time / 60000) * 60000 - time;
+    window.setTimeout(this._updateClock.bind(this), clockUpdateRate);
+  };
+
+  /**
+   * Convert a
+   */
+  proto._timeToString = function epg__timeToString(time) {
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg.js]: Please use l10n_date's API instead of creating this function.

File: tv_apps/tv-epg/js/epg.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+      columnElement.classList.remove('hidden');
+    } else {
+      columnElement.classList.add('hidden');
+    }
+    if (configs.duration) {
+      columnElement.dataset.duration = configs.duration;
+    }
+  };
+
+  proto._updateClock = function epg_updateClock() {
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg.js]: Please use `shared/smart-screen/clock.js`.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=379dd6f0ce73325af3223beffe02c84c300b1760

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a4060ac2b9ceb8b3b9bcc06ef1f40622957f30d0

File: shared/test/unit/mocks/mock_clock.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+'use strict';
```
Comment (huchengtw-moz) [inline shared/test/unit/mocks/mock_clock.js]: Please move to shared/test/unit/mocks/smart-screen/mock_clock.js.

File: shared/test/unit/mocks/mock_smart_bubbles.js [hunk 11]
```diff
@@ -0,0 +1,10 @@
+(function(exports) {
```
Comment (huchengtw-moz) [inline shared/test/unit/mocks/mock_smart_bubbles.js]: Please move to shared/test/unit/mocks/smart-screen/mock_smart_bubbles.js.

File: shared/test/unit/mocks/mock_channel_manager.js [hunk 12]
```diff
@@ -7,9 +7,19 @@
     this.isReady = true;
     this.playingState = {};
 
-    this.getTuner = function() {};
```
Comment (huchengtw-moz) [inline shared/test/unit/mocks/mock_channel_manager.js]: Please move to shared/test/unit/mocks/smart-screen/mock_channel_manager.js.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=59cd644eb112ed54b81836025a62db978734aa97

**Pull request merged**